File: src/utils/ui/uigo/components/button.ts
import { uigo_effects } from "../ui_go_effects";
import { get_sprite_size, is_sprite_hovered, UIGO_ACTION_ID_PRESSED, UIGO_ACTION_ID_RELEASED } from "../ui_go_utils";
import { ObservableVal } from "../../../../utils/observer";
import { v2 } from "../ui_go_util_types";

export function new_uigo_button(go_id: hash, button_click_callback: (url: hash) => void, custum_main_sprite_name = 'sprite', offset_sprite?: v2) {
    const effects = uigo_effects();
    const start_size = get_sprite_size(go_id, custum_main_sprite_name);
    const start_scale = go.get_scale(go_id);
    const hovered = ObservableVal<boolean>(false);
    const is_block = ObservableVal<boolean>(false);

    let hover_effect = true;
    let hover_callback: ((url: hash) => void) | undefined;
    let released_callback: ((url: hash) => void) | undefined;
    let pressed_callback: ((url: hash) => void) | undefined;
    let on_drag_callback: ((url: hash, dx: number, dy: number) => void) | undefined;
    let long_press_callback: ((url: hash) => void) | undefined;

    let _is_btn_pressed = false;
    let _hover_effect_active = false;
    // eslint-disable-next-line prefer-const
    let is_dragged = false;
    let is_long_press = false;

    hovered.subscribe((val) => {

        if (val && _hover_effect_active == false) {
            if (hover_callback != undefined)
                hover_callback(go_id);
            if (hover_effect) {
                _hover_effect_active = true;
                go.set(go_id, 'position.z', 8);
                effects.anim_scale({ url: go_id }, 1.05);
            }
        } else {
            is_long_press = false;
            if (_hover_effect_active) {
                _hover_effect_active = false;
                if (hover_effect) {
                    go.set(go_id, 'position.z', 1);
                    effects.anim_scale({ url: go_id }, start_scale.x);
                }
                _is_btn_pressed = false;
            }
        }

    });

    const is_hovered = (x: number, y: number): boolean => {
        if (!is_block.get()) {
            const result = is_sprite_hovered(go_id, { x: start_size.x, y: start_size.y }, { x, y }, offset_sprite);
            if (hovered.get() != result)
                hovered.set(result);
            return result;
        } else return false;
    };

    const is_pressed = (x: number, y: number, action_id?: string | hash): boolean => {
        if (is_hovered(x, y)) {
            if (action_id == UIGO_ACTION_ID_PRESSED) {
                _is_btn_pressed = true;
                if (pressed_callback != undefined)
                    pressed_callback(go_id);
                //effects.anim_scale({ url: go_id }, 0.95, 0.1, go.EASING_INSINE, () => {
                /* go.animate(go_id, 'scale', go.PLAYBACK_ONCE_FORWARD, vmath.vector3(1.12, 1.12, 1), go.EASING_OUTSINE, 0.2, 0, () => {
                    effects.anim_scale({ url: go_id }, start_scale.x);
                }); */
                //});

            } else if (action_id == UIGO_ACTION_ID_RELEASED) {
                if (released_callback != undefined)
                    released_callback(go_id);

                if (_is_btn_pressed)
                    button_click_callback(go_id);
                hovered.set(false);
                _is_btn_pressed = false;
            }
            return _is_btn_pressed;
        } else {
            return false;
        }
    };

    const trigger_on = () => { button_click_callback(go_id); };

    return {
        url: go_id,
        width: () => start_size.y,
        height: () => start_size.x,
        hovered_effect_enabled: (enabled: boolean) => hover_effect = enabled,
        is_hovered,
        is_pressed,
        pressed_callback,
        hover_callback,
        set_released_callback: (callback: (url: hash) => void) => { released_callback = callback; },
        long_press_callback,
        on_drag_callback,
        trigger_on,
        start_scale,
        is_block,
        is_long_press,
        is_dragged,
        hovered
    };
}

File: app/utils/ui/uigo/components/button.lua
local ____exports = {}
local ____ui_go_effects = require("utils.ui.uigo.ui_go_effects")
local uigo_effects = ____ui_go_effects.uigo_effects
local ____ui_go_utils = require("utils.ui.uigo.ui_go_utils")
local get_sprite_size = ____ui_go_utils.get_sprite_size
local is_sprite_hovered = ____ui_go_utils.is_sprite_hovered
local UIGO_ACTION_ID_PRESSED = ____ui_go_utils.UIGO_ACTION_ID_PRESSED
local UIGO_ACTION_ID_RELEASED = ____ui_go_utils.UIGO_ACTION_ID_RELEASED
local ____observer = require("utils.observer")
local ObservableVal = ____observer.ObservableVal
function ____exports.new_uigo_button(go_id, button_click_callback, custum_main_sprite_name, offset_sprite)
    if custum_main_sprite_name == nil then
        custum_main_sprite_name = "sprite"
    end
    local effects = uigo_effects()
    local start_size = get_sprite_size(go_id, custum_main_sprite_name)
    local start_scale = go.get_scale(go_id)
    local hovered = ObservableVal(false)
    local is_block = ObservableVal(false)
    local hover_effect = true
    local hover_callback
    local released_callback
    local pressed_callback
    local on_drag_callback
    local long_press_callback
    local _is_btn_pressed = false
    local _hover_effect_active = false
    local is_dragged = false
    local is_long_press = false
    hovered.subscribe(function(val)
        if val and _hover_effect_active == false then
            if hover_callback ~= nil then
                hover_callback(go_id)
            end
            if hover_effect then
                _hover_effect_active = true
                go.set(go_id, "position.z", 8)
                effects.anim_scale({url = go_id}, 1.05)
            end
        else
            is_long_press = false
            if _hover_effect_active then
                _hover_effect_active = false
                if hover_effect then
                    go.set(go_id, "position.z", 1)
                    effects.anim_scale({url = go_id}, start_scale.x)
                end
                _is_btn_pressed = false
            end
        end
    end)
    local function is_hovered(x, y)
        if not is_block.get() then
            local result = is_sprite_hovered(go_id, {x = start_size.x, y = start_size.y}, {x = x, y = y}, offset_sprite)
            if hovered.get() ~= result then
                hovered.set(result)
            end
            return result
        else
            return false
        end
    end
    local function is_pressed(x, y, action_id)
        if is_hovered(x, y) then
            if action_id == UIGO_ACTION_ID_PRESSED then
                _is_btn_pressed = true
                if pressed_callback ~= nil then
                    pressed_callback(go_id)
                end
            elseif action_id == UIGO_ACTION_ID_RELEASED then
                if released_callback ~= nil then
                    released_callback(go_id)
                end
                if _is_btn_pressed then
                    button_click_callback(go_id)
                end
                hovered.set(false)
                _is_btn_pressed = false
            end
            return _is_btn_pressed
        else
            return false
        end
    end
    local function trigger_on()
        button_click_callback(go_id)
    end
    return {
        url = go_id,
        width = function() return start_size.y end,
        height = function() return start_size.x end,
        hovered_effect_enabled = function(enabled)
            hover_effect = enabled
            return hover_effect
        end,
        is_hovered = is_hovered,
        is_pressed = is_pressed,
        pressed_callback = pressed_callback,
        hover_callback = hover_callback,
        set_released_callback = function(callback)
            released_callback = callback
        end,
        long_press_callback = long_press_callback,
        on_drag_callback = on_drag_callback,
        trigger_on = trigger_on,
        start_scale = start_scale,
        is_block = is_block,
        is_long_press = is_long_press,
        is_dragged = is_dragged,
        hovered = hovered
    }
end
return ____exports
